lib/services/service-utilities.ts

```diff
@@ -0,0 +1,204 @@
+/*
+ Copyright 2016-2017 Resin.io
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+ http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+ */
+
+import * as _ from 'lodash';
+import TypedError = require('typed-error');
+import * as Promise from 'bluebird';
+import * as bodyParser from 'body-parser';
+import * as express from 'express';
+import { Worker } from '../framework/worker';
+import { WorkerClient } from '../framework/worker-client';
+import { Logger, LogLevel } from '../utils/logger';
+import { ContextAbsentError } from './errors/context-absent';
+import {
+	ServiceAPIHandle, ServiceEmitContext, ServiceEmitRequest, ServiceEmitResponse,
+	ServiceEmitter, ServiceListener, ServiceRegistration,
+} from './service-types';
+import {
+	UtilityServiceEvent, UtilityWorkerEvent
+} from './service-utilities-types';
+
+/**
+ * A utility class to handle a bunch of the repetitive work associated with being a ServiceListener and ServiceEmitter
+ *
+ * This class provides:
+ * - An express app, if asked for.
+ * - A logger.
+ * - An event registration and handling standard.
+ * - Checks for and extracts data from context.
+ * - A standard way of getting contextualised workers.
+ *
+ * In exchange you agree to:
+ * - provide a `emitData` method which takes a ServiceEmitContext and returns a promise.
+ * - provide a `verify` function which should check that incoming data isn't spoofed.
+ * - provide a `serviceName` getter, which because it is based on file path cannot be inherited away.
+ * - provide a `apiHandle` getter, which should return the underlying object that executes the requests.
+ * - enqueue your events with `context` and `event`, as per `UtilityServiceEvent`.
+ */
+export abstract class ServiceUtilities<T> extends WorkerClient<T> implements ServiceListener, ServiceEmitter {
```

Comment:
@pimterry pointed out that this isn't really utilities in the strict sense of the word, as Services inherit from it if they want to use it. I don't really like `AbstractService` as a Service doesn't *have* to inherit from it. `ServiceParent`? `ServiceFramework`? I have no idea! 😺